### PR TITLE
feat: Add a release script

### PR DESCRIPTION
**Pull request by cpitclaudel**
This removes most of the manual steps in the release process, except for the manual CI workflow trigger, which is left for later.

It also changes the way we do release notes: instead of updating a single file (which creates repeated PR conflicts), each release note item goes into a separate text file numbered after the corresponding PR.  The details of this are written up in `docs/news/README` and a link points to that folder in RELEASE_NOTES.md to avoid confusion.

### CODE REVIEW

File: docs/dev/RELEASE.md
Comment (MikaelMayer) [inline docs/dev/RELEASE.md:21]: Can you explicitely say to wait till the tests are finished, and how much time it typically last?

File: docs/dev/news/README.md
Comment (MikaelMayer) [inline docs/dev/news/README.md:12]:
```suggestion
Improved toaster settings prevents Dafny from ever burning toast.
```

File: docs/dev/news/README.md
Comment (MikaelMayer) [inline docs/dev/news/README.md:15]: Can you exactly say how it will be included? I assume that for the above, it will automatically add the line

- fix: Improved toaster settings prevents Dafny from ever burning  (https://github.com/dafny-lang/dafny/pull/1234)

File: Scripts/prepare_release.py
Comment (MikaelMayer) [inline Scripts/prepare_release.py:1]: I have a challenging question for you. How hard would it be to actually convert this code to dafny so that it can be run with `dafny -noVerify ` ?
That way, you wouldn't need to have `assert ends_with('...')` that I see in the code.
Another advantage would be that we would be able to execute tests that would greatly enhance the maintenance process. There are tons of regexps that might break that would be easy to fix if only we saw what the typical input and typical output should look like.

**Review by robin-aws — CHANGES_REQUESTED** [review]
On a quick glance this looks awesome, very much appreciate the contribution to streamlining the release process!

I will find time soon to review properly, but my high-level strong preference is to not require doing any work on a local machine (fully scripted or not, but to find a way for it to happen in a GitHub Action workflow. There is always risk for local development state to corrupt the release process, sometimes very subtly.

File: Scripts/prepare_release.py
Comment (MikaelMayer) [inline Scripts/prepare_release.py:107]: I would prefer very much to have complete and explicit names, so that it does not feel like a compiled code :-)

```suggestion
    version_number: str # Main version number (1.2.3)
    release_date: date # Release date
    version_number_suffix: str # Optional marker ("alpha")
```

File: Scripts/prepare_release.py
Comment (MikaelMayer) [inline Scripts/prepare_release.py:103]:
```suggestion
    VERSION_NUMBER_REGEX = re.compile("^(?P<prefix>[0-9]+[.][0-9]+[.][0-9]+)(?P<suffix>-.+)?$")
```

**Review by MikaelMayer — CHANGES_REQUESTED** [review]
Yay, wonderful time-saver script, which will help us release more often I hope.

**Comment by MikaelMayer** [comment]
> I will find time soon to review properly, but my high-level strong preference is to not require doing any work on a local machine (fully scripted or not, but to find a way for it to happen in a GitHub Action workflow. There is always risk for local development state to corrupt the release process, sometimes very subtly.

I think that, in a first step, mimicking what we used to do manually locally anyway is a good compromise. I tested in on my Windows machine and this script worked.
Later, we can always try to automate more the process and make it purely online, but I think that would be an improvement, not a blocker over this PR.

File: docs/dev/news/README.md
Comment (cpitclaudel) [reply docs/dev/news/README.md:15]: It actually becomes 

```markdown
## Bug fixes

- Improved toaster settings prevent Dafny from ever burning toast.
  (https://github.com/dafny-lang/dafny/pull/1234)
```

I don't think the README should say this, though.

File: Scripts/prepare_release.py
Comment (cpitclaudel) [reply Scripts/prepare_release.py:1]: It would be anywhere between a huge pain and impossible, because of the range of libraries used: XML parsing, regular expressions, HTTP requests, dates and times, text manipulation, file system functions, quoting, etc (but not JSON parsing, because we do have that! ^^)

I can remove these assertions if you'd like; they'd just become requires in Dafny.  They are here to make sure that the messages are readable.

Python has better support for tests than Dafny; I can add tests if you want (but there's decreasing returns to spending days working on a script that runs once a month ^^)

File: Scripts/prepare_release.py
Comment (cpitclaudel) [reply Scripts/prepare_release.py:103]: Renamed to `VERSION_NUMBER_PATTERN`, but I find it much harder to read (because the long names hide the rest of the code)

**Comment by cpitclaudel** [comment]
> Fair enough, I just need the deeper review to verify we're not doing more locally than before. :)

Do you want to block this, or are you OK with Mikael's review?

File: docs/dev/news/README.md
Comment (MikaelMayer) [reply docs/dev/news/README.md:15]: I agree this is an implementation detail, but it's a detail important enough so that I seriously think it's worth mentioning it.
Otherwise, we might get developers down the road who will want to prefix their subjects with "fix:" or "feat:" or want to write something on two lines, or they will want to include a custom URL, or they won't know if they have to put a capital or not on the first letter, or if they need to include a space at the end.... all of this because the transformation is only `provide`d but not `reveal`ed :-)

And we should get rid of the need to add "??". By default, the URL should be appended automatically. I don't see a clear use case where a developer would want to include this link anywhere in the subject link.

**Comment by cpitclaudel** [comment]
> I will find time soon to review properly, but my high-level strong preference is to not require doing any work on a local machine 

Agreed, and this script can run from CI as well.  The only difficulty would be creating a bot account on github that is allowed to push to the repo.

But since this sort of script invariably takes a few iterations to refine and since it's much easier to debug locally than on CI, I think it's best to start locally.

File: Scripts/prepare_release.py
Comment (MikaelMayer) [reply Scripts/prepare_release.py:1]: We will probably want to run this script once a week. I'm not saying we have to convert this script to Dafny, but if it's challenging to write such a script in Dafny, then maybe it's a yellow flag as per how difficult it is to integrate with an existing environment.
Dafny compiles to python as well, so it should not be hard to have wrapper libraries.

File: Scripts/prepare_release.py
Comment (MikaelMayer) [reply Scripts/prepare_release.py]: In the same answer spirit, it's not clear with "prefix" or "suffix" what of the version numbers are included in which :-)

Maybe the following could reach a consensus?
`prefix` => `major_minor_patch`
`suffix` stays the same.

File: Scripts/prepare_release.py
Comment (MikaelMayer) [reply Scripts/prepare_release.py]: Thanks for the renaming, I think it's appropriate. How does the the long name hide the rest of the code? I'm not sure to understand this part.

File: Scripts/prepare_release.py
Comment (cpitclaudel) [reply Scripts/prepare_release.py:1]: It's hundreds of lines of wrapper code / dozens of wrappers; I don't think Dafny is ready to be a scripting language (and that's OK)

File: docs/dev/news/README.md
Comment (cpitclaudel) [reply docs/dev/news/README.md]: OK, done.  I also added a test to make sure that the format was correct.

> And we should get rid of the need to add "??". By default, the URL should be appended automatically. 

`??` was a typo, you spotted it and I already accepted your patch.  Did I miss something else?

File: Scripts/prepare_release.py
Comment (cpitclaudel) [reply Scripts/prepare_release.py]: OK, I renamed `prefix` to `main` and `suffix` to `identifier` (the name used for this in semantic versioning)

File: Scripts/prepare_release.py
Comment (cpitclaudel) [reply Scripts/prepare_release.py]: Very long function and variable names make code harder to read, because the relevant bits of syntax / control end up further from each other, at least in my experience.

File: docs/dev/news/README.md
Comment (MikaelMayer) [reply docs/dev/news/README.md]: Oh I probably missed your fix. Thanks :-)

File: Scripts/prepare_release.py
Comment (MikaelMayer) [inline Scripts/prepare_release.py:119]: I don't think you want to add a newline between the contents and the link. At least in GitHub, where the release nodes can be seen as a formatted markdown file, even though the link seems long, it will actually look like this: #1239

File: Scripts/prepare_release.py
Comment (MikaelMayer) [inline Scripts/prepare_release.py:80]: Thanks for the testing code ! It helps a lot.

```suggestion
    - Improved toaster settings.
      Dafny will not burn toast again (https://github.com/dafny-lang/dafny/pull/1234)
```
```suggestion
    - Improved toaster settings.
      Dafny will not burn toast again.
      (https://github.com/dafny-lang/dafny/pull/1234)
```
See comment below for the formatting.
Besides the extra newline that is not necessarily ([have a look at here to see how release notes are currently formatted](https://github.com/dafny-lang/dafny/releases))I would also add that you would now be forcing users (and me) to do something that we don't want: Add a period before the parentheses.
So, in your example, and in your README, could you be specific about this point? Alternatively (and that would be my favorite method), you could detect the last character of the trimmed string, and if it's a period, you'd remove it.

**Review by MikaelMayer — CHANGES_REQUESTED** [review]
Still one more cosmetic change to make it easier for everyone and ensure we are markdown-friendly.

File: Scripts/prepare_release.py
Comment (cpitclaudel) [reply Scripts/prepare_release.py]: Done: I've removed the newline.  I find that it makes reading the release notes harder, but I won't fight about this.

The existing `RELEASE_NOTES` are inconsistent on whether to use a period,  so I've changed the code to add a period in all cases.  The version above with one period for the first sentence and no period for the second doesn't look right at all to me; do you like it better?

Comment (MikaelMayer) [reply Scripts/prepare_release.py]: >  I find that it makes reading the release notes harder

I won't argue since you implemented my change, thanks :-) but I'm just curious.
Can you please clarify, is it because you read the source of release notes instead of in a browser? Or is the browser rendering itself harder to read? To me, the first image here:
image
is much nicer than the alternative that you might find more appealing based on your script:
image
I find that the second one breaks the flow, it feels like every link is important when the only thing I would like out of release notes is to read them in diagonal.

Regarding two lines: I think we should not even consider the possibility that a feat/fix line in a release notes should have two sentences. I know you added support for it, but it's not something we have done in the past and I don't see a reason why we should embrace those. So if we were to add paragraphs, I'd agree with you, we should end every sentence with a full stop, and make the link its own period-less sentence after that.
However, for a single phrase, with the comment in parentheses containing the link being part of the phrase, there should be no period between the phrase and the parentheses. If there was a period, it should come afterwards.

Now, one more principle to consider: We use bullet points to express each individual release notes, and many bullet points are not complete sentences but just fragments. In this case I think it's [conventional wisdom](https://www.instructionalsolutions.com/blog/bulleted-list-punctuation) to omit the period at the end of the sentence (after the parentheses containing the link).

File: docs/dev/news/README.md
Comment (MikaelMayer) [inline docs/dev/news/README.md:13]: If you don't like my comment above and still find that multi-line release notes are relevant, like, we want to output more user-centric release notes that not only have a title, but a small example or explanation:

```suggestion
Added support for :==, like this:
    method Test(x: int, y: int) {
      var x := 1;
      x :== 1; // Assertion succeeds.
      x :== 2; // Assertion fails.
    }
```

Then, in this case, if it's multiline, you can add a newline (no automatically inserted period) and put the link in one more link. But if someone writes only one line, then if there is a period, then you move it after the (link) that is inserted in the sentence.

File: Scripts/prepare_release.py
Comment (cpitclaudel) [reply Scripts/prepare_release.py]: Good point.  I implemented your suggestion below: no auto punctuation fix and no newline before link, unless there's already multiple lines in the entry.

We can refine the rendering further, probably when we have more data (before the next release)

File: docs/dev/news/README.md
Comment (MikaelMayer) [inline docs/dev/news/README.md:24]:
```suggestion
At release time, these two files become bullet points in a new section of `RELEASE_NOTES.md` with links to the corresponding PRs, like this:
```

File: docs/dev/RELEASE.md
Comment (robin-aws) [inline docs/dev/RELEASE.md:40]: I don't love the idea of releasing from a branch before merging it to mainline. Could we flip the order of steps to do this first, even if it means customizing the GHA that looks for the deep tests? Are we just trying to save the extra delay on waiting for the CI?

File: RELEASE_NOTES.md
Comment (robin-aws) [inline RELEASE_NOTES.md:3]: I'd add a stronger comment that this file is no longer used as part of the release process.

File: docs/dev/RELEASE.md
Comment (robin-aws) [inline docs/dev/RELEASE.md:8]: Could we have the script infer the correct next version number based on whether there are any new `feat` or `feat!` news items? Or at least verify you selected the right version number.

**Review by robin-aws — CHANGES_REQUESTED** [review]
I just reviewed the new workflow and did not look closely at the implementation, will leave that to other reviewers.

On the one hand I definitely love the new approach. On the other I do feel like this is increasing the work that has to happen on a developer's machine, since in the old way we were manually creating the release notes incrementally in each PR. Let me think a bit more on how to avoid that.

File: docs/dev/RELEASE.md
Comment (cpitclaudel) [reply docs/dev/RELEASE.md:40]: Releasing from master forces a freeze to avoid concurrent commits; we've been bitten by that before.  What issue do you see with releasing from a branch?  Is it the security aspect? (I think we already allow releasing by uploading a custom-packaged zip, so I haven't worried about that part)

File: RELEASE_NOTES.md
Comment (cpitclaudel) [reply RELEASE_NOTES.md:3]: This is a user-facing file, though, so should we add this type of comments there?
Maybe as a comment that doesn't get rendered?

File: docs/dev/RELEASE.md
Comment (cpitclaudel) [reply docs/dev/RELEASE.md:8]: Sure, do you think it's worth it given that we plan to revise our version policy soon?

File: docs/dev/RELEASE.md
Comment (cpitclaudel) [reply docs/dev/RELEASE.md:19]: I would prefer to keep it here.  It's good to be reminded of the issue every time we go through these instructions.  But le me know if you disagree and I'll move it.

File: docs/dev/RELEASE.md
Comment (MikaelMayer) [reply docs/dev/RELEASE.md:40]: I think it's not only pretty common to release from a branch before merging it (that way, it's even possible to cherry-pick bug fixes independently without introducing new features), but in our case, as Clement said, we already encountered the case when we wanted to release after merging with master; however, there was one more commit on master since we started to prepare the release, and this commit was merged unintentionally with the release.
So although I had wished that releases are made from the history of master branches, in practice, this is dangerous.

File: RELEASE_NOTES.md
Comment (MikaelMayer) [reply RELEASE_NOTES.md:3]: I don't think we should add a comment about the release process here, but I like the idea of also redirecting users to the folder where the fragments are so that they can see what to expect in the next release.

File: docs/dev/RELEASE.md
Comment (MikaelMayer) [reply docs/dev/RELEASE.md:8]: I'm afraid we would loose the ability to bump the major version number with such an auto-detect 😮 
My best suggestion is that:
- The script detect feat or feat! extensions
- If there are, it suggests to increment the minor version number, and if the user presses ENTER, this is it.
- If there are none (only fixes), it suggests to increment the patch version number, and if the user presses ENTER, this is it
- But if the user decides to put another version number, it takes it and continue.

**Review by MikaelMayer — COMMENTED** [review]
Removing my "request change"

**Review by MikaelMayer — DISMISSED** [review]
Commenting does not remove the "Request change". Let me approve it.

File: docs/dev/RELEASE.md
Comment (cpitclaudel) [reply docs/dev/RELEASE.md:8]: This is a lot of work for saving the user from typing 3 numbers ^^  I vote for leaving this for future improvement

File: RELEASE_NOTES.md
Comment (robin-aws) [reply RELEASE_NOTES.md:3]: Ah I missed the fact that the script is still updating this file, I thought it was going to be frozen in this state forever more. Nevermind. :)

File: docs/dev/RELEASE.md
Comment (robin-aws) [reply docs/dev/RELEASE.md:40]: Okay that's a better reason, I'm happy with this.

**Review by robin-aws — DISMISSED** [review]

**Review by robin-aws — DISMISSED** [review]

**Review by robin-aws — APPROVED** [review]

**Pull request merged**
